### PR TITLE
Improve benchmarks

### PR DESCRIPTION
**Pull request by yguclu**
- Don't change arguments between successive calls to a test function
- Clean up functions' setup
- Allow running any given test manually without copying setup instructions from the script `run_benchmarks.py`
- Make sure that every test returns a number or an array, which can be checked for correctness against the pure-Python version
- Increase the run-time of the accelerated functions to at least a good fraction of a second

### CODE REVIEW

**Review by EmilyBourne — APPROVED** [review]

**Review by EmilyBourne — APPROVED** [review]

File: benchmarks/run_benchmarks.py [hunk 1]
```diff
@@ -86,34 +86,34 @@
...
+        '''import numpy as np; nx=3001; nt=3000; c=1.; dt=0.000000003;
         dx = 2 / (nx-1);
         u0 = np.ones(nx);
         u0[int(.5 / dx):int(1 / dx + 1)] = 2;''',
         'u = linearconv_1d(u0, nt, dt, dx, c)'),
     TestInfo('FD - NL Convection',
         'nonlinearconv_1d_mod.py',
         ['nonlinearconv_1d'],
-        '''import numpy as np; nx = 2001; nt=2000; c=1.; dt=0.00035;
+        '''import numpy as np; nx = 3001; nt=3000; c=1.; dt=0.000000003;
```
Comment (yguclu) [inline benchmarks/run_benchmarks.py:116]: Not sure why you would want to decrease the time step `dt` so drastically. All we usually need is that a Courant condition is satisfied, which is roughly `dt < dx / c` (and I think that `dx = 1 / nx` in our case, so the condition becomes `dt < 1 / nx`).

File: benchmarks/tests/bellman_ford_mod.py [hunk 2]
```diff
@@ -53,19 +53,31 @@ def bellman_ford_test ( ):
...
+    weights = zeros (e_num, dtype = float )
+    idx = 0
 
-    e_weight = array( (-3.0,  6.0, -4.0, -1.0,  4.0, \
-                       -2.0,  2.0,  8.0, -3.0,  3.0 ) )
+    for i in  range( v_num ):
+        for j in range( v_num ):
+            if i > j:
+                e[0][idx] = i
+                e[1][idx] = j
```
Comment (EmilyBourne) [inline benchmarks/tests/bellman_ford_mod.py:69]: For NumPy arrays it is usual (and more efficient) to index like:

```suggestion
                e[0, idx] = i
                e[1, idx] = j
```

File: benchmarks/tests/numba_bellman_ford_mod.py [hunk 3]
```diff
@@ -54,19 +55,29 @@ def bellman_ford_test ( ):
...
+    weights = zeros (e_num, dtype = float )
+    idx = 0
 
-    e = array( (( 1, 4, 1, 2, 4, 2, 5, 3, 5, 3 ), \
-                ( 0, 1, 2, 4, 0, 5, 0, 2, 3, 0 )) )
+    for i in  range( v_num ):
+        for j in range( v_num ):
+            if i > j:
+                e[0][idx] = i
+                e[1][idx] = j
```
Comment (EmilyBourne) [inline benchmarks/tests/numba_bellman_ford_mod.py:69]: See above

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: benchmarks/run_benchmarks.py (hunk 1 above)
Comment (jalalium) [reply benchmarks/run_benchmarks.py:116]: I just tweaked the input to make the test slower without taking into the consideration its theoritical significance. I will revert these changes

Comment (yguclu) [reply benchmarks/run_benchmarks.py:116]: Does `dt` effect the runtime?

Comment (jalalium) [reply benchmarks/run_benchmarks.py]: if one increases nx and nt, dx gets larger and the values get too large too quickly so dt needs to be decreased to keep the values from blowing up

File: benchmarks/tests/bellman_ford_mod.py (hunk 2 above)
Comment (jalalium) [reply benchmarks/tests/bellman_ford_mod.py]: Done

File: benchmarks/tests/numba_bellman_ford_mod.py (hunk 3 above)
Comment (jalalium) [reply benchmarks/tests/numba_bellman_ford_mod.py]: Done

File: benchmarks/run_benchmarks.py (hunk 1 above)
Comment (yguclu) [reply benchmarks/run_benchmarks.py]: I've just realized that this is the non-linear convection case, so the variable `c` is not needed at all.

Comment (yguclu) [reply benchmarks/run_benchmarks.py]: Just to clarify, `nx` is the number of points in the domain of length `2`, so the cell size is `dx = 2 / (nx - 1)`. The propagation speed for a the non-linear convection case is the solution `u` itself, which has a maximum of 2 in the initial conditions. The explicit Euler time stepping scheme must satisfy the stability limit `dt < dx / max(u)`(the Courant condition mentioned above). If the time step size `dt` is chosen too large the scheme is unstable, this is why you observe a growth in the amplitude of the solution.

Comment (yguclu) [reply benchmarks/run_benchmarks.py]: By the way, the model that is solved in this example is the 1D inviscid Burgers' equation, which can develop shocks in the solution after a certain amount of time (see [Wikipedia](https://en.wikipedia.org/wiki/Burgers%27_equation) for example). This should not be a problem as we are using one-sided finite differences (a.k.a. "upwinding") to compute the derivative `du/dx`.

**Comment by yguclu** [comment]
@jalalium Please avoid force-pushing to the branches on GitHub, in order to avoid conflicts if somebody else is working on the same branch.

File: benchmarks/run_benchmarks.py (hunk 1 above)
Comment (jalalium) [reply benchmarks/run_benchmarks.py]: Thank you very much for the clarification!

**Comment by jalalium** [comment]
> @jalalium Please avoid force-pushing to the branches on GitHub, in order to avoid conflicts if somebody else is working on the same branch.

My apologies, it will not happen again.

File: benchmarks/tests/md_mod.py [hunk 4]
```diff
@@ -57,8 +57,6 @@ def compute ( p_num: int, d_num: int, pos: 'double[:,:]', vel: 'double[:,:]',
 
     kinetic = 0.5 * mass * kinetic
 
-    return potential, kinetic
```
Comment (EmilyBourne) [inline benchmarks/tests/md_mod.py:60]: This is removed. But the arrays are still calculated. Shouldn't the calculation also be removed given that we don't use the result?

File: benchmarks/tests/numba_md_mod.py [hunk 5]
```diff
@@ -93,11 +93,10 @@ def r8mat_uniform_ab ( r: 'double[:,:]', m: int, n: int, a: float, b: float, see
         seed = seed + i4_huge
 
     elif ( seed == 0 ):
-#        print ( '' ) # TODO error in Pyccel
+        print ( '' )
         print ( 'R8MAT_UNIFORM_AB - Fatal error!' )
         print ( '  Input SEED = 0!' )
-        # TODO must stop here
-#        sys.exit ( 'R8MAT_UNIFORM_AB - Fatal error!' )
+        sys.exit ( 'R8MAT_UNIFORM_AB - Fatal error!' )
```
Comment (EmilyBourne) [inline benchmarks/tests/numba_md_mod.py:99]: This change matches what pyccel can do, but doesn't really make sense in the benchmarks

Comment (EmilyBourne) [reply benchmarks/tests/numba_md_mod.py]: @yguclu Proposed solution:  f0d42c7

File: benchmarks/tests/md_mod.py (hunk 4 above)
Comment (EmilyBourne) [reply benchmarks/tests/md_mod.py:60]: @yguclu Proposed solution to remove code whose result is ignored : bca47a4

Comment (yguclu) [reply benchmarks/tests/md_mod.py:60]: I did not notice this removal. Every test in our benchmark suite should return some kind of output, in order to allow the user/developer to check the correctness of the accelerated code.

Comment (yguclu) [reply benchmarks/tests/md_mod.py:60]: @jalalium Didn't we decide that the test should return the total potential and kinetic energy at the end of the simulation?

Comment (EmilyBourne) [reply benchmarks/tests/md_mod.py:60]: The md test does return the potential and kinetic energy at the end of the simulation. The `compute` function is called at each timestep.

Previously the energy was calculated, returned and ignored. It was changed so it was calculated but not returned, then recalculated in `test_md` after the simulation and the final energies are returned.

This commit removes the calculation inside the timestep whose result is unused

Comment (EmilyBourne) [reply benchmarks/tests/md_mod.py:60]: That said, in order to return the energy from `test_md` it seems that the compute function has simply been duplicated. Either the force update should be removed from `test_md` or we should revert to the original organisation but also return the energy from `md`

Comment (yguclu) [reply benchmarks/tests/md_mod.py:60]: If you see a solution for avoiding useless calculations at each time step and code duplication, please go ahead!

Comment (EmilyBourne) [reply benchmarks/tests/md_mod.py:60]: That is what I did in , I also added 0ac85cb to further reduce code duplication. But there is still some duplication between `compute` and `test_md`.

I have now added 7a9951b which reorganises the file somewhat, but I think this is the cleanest option

Comment (yguclu) [reply benchmarks/tests/md_mod.py:60]: 👍 It looks good indeed!

Comment (jalalium) [reply benchmarks/tests/md_mod.py:60]: > @jalalium Didn't we decide that the test should return the total potential and kinetic energy at the end of the simulation?

It does in the main test function, returning it from the `calculate` function itself raised an error in `Fortran`

Comment (EmilyBourne) [reply benchmarks/tests/md_mod.py:60]: What kind of error? Does the new file compile?

Comment (jalalium) [reply benchmarks/tests/md_mod.py:60]: Yes, your last update works fine in `Fortran` and `C` and gives the correct output

Comment (jalalium) [reply benchmarks/tests/md_mod.py:60]: but this version does not work with `Fortran`

```


# coding: utf-8
#------------------------------------------------------------------------------------------#
# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
#------------------------------------------------------------------------------------------#
"""
Functions for running a small molecular dynamics simulation. The code is adapted from examples written by [J. Burkardt](https://people.sc.fsu.edu/~jburkardt/py_src/py_src.html)
To be accelerated with pyccel or pythran
"""
import sys
from numpy import zeros
from numpy import sqrt
from numpy import pi
from numpy import sin

# ================================================================
def compute ( p_num: int, d_num: int, pos: 'double[:,:]', vel: 'double[:,:]',
             mass: float, force: 'double[:,:]' ):
    """ Calculate the energy and forces associated with the current configuration
    """

    rij = zeros ( d_num )

    potential = 0.0

    for i in range ( 0, p_num ):
        #
        #  Compute the potential energy and forces.
        #
        for j in range ( 0, p_num ):
            if ( i != j ):
                #  Compute RIJ, the displacement vector.
                for k in range ( 0, d_num ):
                    rij[k] = pos[k,i] - pos[k,j]

                #  Compute D and D2, a distance and a truncated distance.
                d = 0.0
                for k in range ( 0, d_num ):
                    d = d + rij[k] ** 2

                d = sqrt ( d )
                d2 = min ( d, pi / 2.0 )

                #  Attribute half of the total potential energy to particle J.
                potential = potential + 0.5 * sin ( d2 ) * sin ( d2 )

                #  Add particle J's contribution to the force on particle I.
                for k in range ( 0, d_num ):
                    force[k,i] = force[k,i] - rij[k] * sin ( 2.0 * d2 ) / d
    #
    #  Compute the kinetic energy.
    #
    kinetic = 0.0
    for k in range ( 0, d_num ):
        for j in range ( 0, p_num ):
            kinetic = kinetic + vel[k,j] ** 2

    kinetic = 0.5 * mass * kinetic

    return potential, kinetic

# ================================================================
def update ( p_num: int, d_num: int, pos: 'double[:,:]', vel: 'double[:,:]',
            force: 'double[:,:]', acc: 'double[:,:]', mass: float, dt: float ):
    """ Update the position, velocity and force of the particles
    """

    rmass = 1.0 / mass
    #
    #  Update positions.
    #
    pos += vel * dt + 0.5 * acc * dt * dt
    #
    #  Update velocities.
    #
    vel += 0.5 * dt * ( force * rmass + acc )
    #
    #  Update accelerations.
    #
    acc[:] = force * rmass

# ================================================================
def r8mat_uniform_ab ( r: 'double[:,:]', m: int, n: int, a: float, b: float, seed: int ):
    """ Fill r with random numbers with a uniform distribution
    """

    i4_huge = 2147483647

    if ( seed < 0 ):
        seed = seed + i4_huge

    elif ( seed == 0 ):
        print ( '' )
        print ( 'R8MAT_UNIFORM_AB - Fatal error!' )
        print ( '  Input SEED = 0!' )
        sys.exit ( 'R8MAT_UNIFORM_AB - Fatal error!' )

    elif ( seed > 0 ):

        for j in range ( 0, n ):
            for i in range ( 0, m ):

                k = ( seed // 127773 )

                seed = 16807 * ( seed - k * 127773 ) - k * 2836

                seed = ( seed % i4_huge )

                if ( seed < 0 ):
                    seed = seed + i4_huge

                r[i,j] = a + ( b - a ) * seed * 4.656612875E-10

    return seed

# ================================================================
def initialize ( pos: 'double[:,:]', p_num: int, d_num: int ):
    """ Initialise the positions of the particles
    """
    #  Positions.
    seed = 123456789
    seed = r8mat_uniform_ab ( pos, d_num, p_num, 0.0, 10.0, seed )

# ================================================================
def md (d_num: int, p_num: int, step_num: int, dt: float,
        vel: 'double[:,:]', acc: 'double[:,:]',
        force: 'double[:,:]', pos: 'double[:,:]'):
    """ Run molecular dynamics simulation
    """

    mass = 1.0

    initialize ( pos, p_num, d_num )
    compute ( p_num, d_num, pos, vel, mass, force )
    
    for _ in range ( step_num - 1):
        update ( p_num, d_num, pos, vel, force, acc, mass, dt )
        compute ( p_num, d_num, pos, vel, mass, force )

    update ( p_num, d_num, pos, vel, force, acc, mass, dt )
    return compute ( p_num, d_num, pos, vel, mass, force )

# ================================================================
# pythran export test_md(int,int,int,float)
def test_md ( ):
    """ Run molecular dynamics test
    """

    #  Input
    d_num = 3
    p_num = 100
    step_num = 5000
    dt = 0.1
    #  Velocities.
    vel = zeros ( ( d_num, p_num ) )
    #  Accelerations.
    acc = zeros ( ( d_num, p_num ) )
    # Forces
    force = zeros ( ( d_num, p_num ) )
    # Positions
    pos = zeros ( ( d_num, p_num ) )

    return md(d_num, p_num, step_num, dt, vel, acc, force, pos)


if __name__ == '__main__':
    print(test_md())
```

It gives the following error:

```

  178 |       allocate(Out_0001_ptr(0:size(Out_0001_ptr, kind=i64) - 1_i64))
      |                           1
Error: ‘out_0001_ptr’ must not appear in the array specification at (1) in the same ALLOCATE statement where it is itself allocated
```

Comment (EmilyBourne) [reply benchmarks/tests/md_mod.py:60]: I have opened https://github.com/pyccel/pyccel/issues/1321

**Pull request merged**